### PR TITLE
Publish charmverse proposal comments to lens

### PR DESCRIPTION
**Pull request by Devorein**
### <samp>🤖 Generated by Copilot at b9be652</samp>

This pull request adds the feature to publish proposals to the lens app, which is a decentralized platform for sharing and discussing documents. It also adds the feature to create lens comments for published proposals and to scroll to a specific comment. It updates the types, functions, components, hooks, and api routes related to proposals, comments, and lens publications. It also adds new files for creating comment publications, uploading to arweave, and updating proposal lens properties.

### WHY
1. Added publish to lens toggle for proposal
2. Added lens post link in proposal properties
3. Publish comment to lens

### CODE REVIEW

File: charmClient/apis/proposalsApi.ts [hunk 1]
```diff
@@ -8,12 +8,17 @@ import type { ArchiveProposalRequest } from 'lib/proposal/archiveProposal';
...
 import type { UpdateProposalRequest } from 'lib/proposal/updateProposal';
+import type { UpdateProposalLensPropertiesRequest } from 'lib/proposal/updateProposalLensProperties';
 
 export class ProposalsApi {
   updateProposal({ proposalId, ...rest }: UpdateProposalRequest) {
     return http.PUT(`/api/proposals/${proposalId}`, rest);
   }
 
+  updateProposalLensProperties({ proposalId, ...rest }: UpdateProposalLensPropertiesRequest) {
+    return http.PUT(`/api/proposals/${proposalId}/update-lens-properties`, rest);
```
Comment (mattcasey) [inline charmClient/apis/proposalsApi.ts:19]: please start using hooks, eg. charmClient/hooks/proposals.ts. This is actually the one api i've already started converting to hooks

File: components/[pageId]/DocumentPage/components/ProposalProperties.tsx [hunk 2]
```diff
@@ -81,8 +82,15 @@ export function ProposalProperties({
   async function updateProposalStatus(newStatus: ProposalStatus) {
     if (proposal && newStatus !== proposal.status) {
       await charmClient.proposals.updateStatus(proposal.id, newStatus);
-      if (newStatus === 'discussion' && proposalPage) {
-        await createPost(proposalPage);
+      // TODO: Handle when proposal is published to lens
+      if (newStatus === 'discussion' && proposalPage && proposal.publishToLens) {
+        const createdLensPost = await createPost(proposalPage);
+        if (createdLensPost) {
+          await charmClient.proposals.updateProposalLensProperties({
```
Comment (mattcasey) [inline components/[pageId]/DocumentPage/components/ProposalProperties.tsx:89]: it seems like this should be part of the useLensProfile post, and you could just export a method called 'createOrUpdatePost'

File: components/settings/account/hooks/useLensProfile.tsx [hunk 3]
```diff
@@ -1,22 +1,33 @@
...
 import { switchActiveNetwork } from 'lib/blockchain/switchNetwork';
+import { createCommentPublication } from 'lib/lens/createCommentPublication';
 import { createPostPublication } from 'lib/lens/createPostPublication';
 import { lensClient } from 'lib/lens/lensClient';
 import type { PageWithContent } from 'lib/pages';
+import type { PageContent } from 'lib/prosemirror/interfaces';
 import { generateMarkdown } from 'lib/prosemirror/plugins/markdown/generateMarkdown';
 
-const CHAIN: Blockchain = isProdEnv ? 'POLYGON' : 'MUMBAI';
+const CHAIN: Blockchain = !isDevEnv ? 'MUMBAI' : 'POLYGON';
```
Comment (mattcasey) [inline components/settings/account/hooks/useLensProfile.tsx:30]: @Devorein I think this needs to be flipped before merge?

File: lib/lens/lensClient.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@ import { isDevEnv } from 'config/constants';
 const isServer = typeof window === 'undefined';
 
 const lensClient = new LensClient({
-  environment: isDevEnv ? development : production,
+  environment: !isDevEnv ? development : production,
```
Comment (mattcasey) [inline lib/lens/lensClient.ts:8]: another test ?

File: lib/lens/uploadToArweave.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { POST } from 'adapters/http';
+
+export const uploadToArweave = async (data: any): Promise<string> => {
+  const response = await POST<string>('https://metadata.lenster.xyz', data, {
+    headers: { 'Content-Type': 'application/json' },
+    // remove credentials to bypass CORS error
+    credentials: 'omit'
+  });
+  // Lenster response header content type is text/plain;charset=UTF-8, so we need to json parse it manually
+  const { id } = JSON.parse(response);
```
Comment (mattcasey) [inline lib/lens/uploadToArweave.ts:10]: should try/catch

File: components/proposals/components/ProposalProperties/ProposalProperties.tsx [hunk 6]
```diff
@@ -157,6 +162,12 @@ export function ProposalProperties({
     }
   }
 
+  useEffect(() => {
+    setProposalFormInputs({
+      publishToLens: proposalFormInputs?.publishToLens ?? !!user?.publishToLensDefault
```
Comment (mattcasey) [inline components/proposals/components/ProposalProperties/ProposalProperties.tsx:167]: This component is not responsible for defaults - publishToLens should be defined in the components in DocumentPage/ProposalProperties and NewProposalPage

File: components/proposals/components/ProposalProperties/ProposalProperties.tsx [hunk 7]
```diff
@@ -397,6 +408,56 @@ export function ProposalProperties({
               />
             </Box>
           </Box>
+
+          {/* Publish to lens toggle */}
+          {user && (
```
Comment (mattcasey) [inline components/proposals/components/ProposalProperties/ProposalProperties.tsx:413]: it doesn't seem like u need to wait for user to load to show this.

File: components/settings/account/hooks/useLensProfile.tsx [hunk 8]
```diff
@@ -88,33 +112,61 @@ export function useLensProfile() {
...
-        contentText: `Proposal ${proposal.title} from ${space.name} is now open for feedback.\n\n${
-          isMarkdownContentOverflowing
-            ? `${markdownContent.slice(0, LENS_PROPOSAL_PUBLICATION_LENGTH)}...`
-            : markdownContent
-        }`,
-        proposalLink: `https://app.charmverse.io/${space.domain}/${proposal.path}`,
-        lensProfile
-      });
-      if (postPublication.data.isFailure()) {
+      let publicationResponse: {
```
Comment (mattcasey) [inline components/settings/account/hooks/useLensProfile.tsx:128]: suggestion: define this type outside the function, its a very complicated type

**Review by mattcasey — CHANGES_REQUESTED** [review]
please double-check the variables set on env. And I think it would be better if you move the check for user preferences outside of the base properties component. It's not supposed to be opinionated about defaults

File: components/settings/account/hooks/useLensProfile.tsx (hunk 3 above)
Comment (Devorein) [reply components/settings/account/hooks/useLensProfile.tsx:30]: Right. Will revert it before merge.

File: components/[pageId]/DocumentPage/components/ProposalProperties.tsx (hunk 2 above)
Comment (Devorein) [reply components/[pageId]/DocumentPage/components/ProposalProperties.tsx:89]: I've moved it inside of the createPost function from lens. These two are kindof atomic so doesn't make sense to separate them

File: lib/lens/lensClient.ts (hunk 4 above)
Comment (Devorein) [reply lib/lens/lensClient.ts:8]: right. bookmarked all of them to switch before deploy

File: charmClient/apis/proposalsApi.ts (hunk 1 above)
Comment (Devorein) [reply charmClient/apis/proposalsApi.ts:19]: Updated. really love the new design. It removes putting everything inside a single all-purpose hook. Divided up useLensProfile to `useLensProfile` & `useLensPulication` hooks.

File: lib/lens/uploadToArweave.ts (hunk 5 above)
Comment (Devorein) [reply lib/lens/uploadToArweave.ts:10]: added

File: components/proposals/components/ProposalProperties/ProposalProperties.tsx (hunk 7 above)
Comment (Devorein) [reply components/proposals/components/ProposalProperties/ProposalProperties.tsx:413]: You're right. Removed the user condition.

File: components/settings/account/hooks/useLensProfile.tsx (hunk 8 above)
Comment (Devorein) [reply components/settings/account/hooks/useLensProfile.tsx:128]: updated

**Pull request merged**